### PR TITLE
3D beta rendering support for SceneEntity `triangles`

### LINKED ISSUE #4318: 3D beta rendering support for SceneEntity `triangles`

### PR DESCRIPTION
**Pull request by snosenzo**
**User-Facing Changes**
 - TriangleList scene entity primitive enabled for foxglove schema

**Description**
 - enable triangle list scene entity primitive
 - ~NOTE: upon second playthrough it makes the tab freeze at some time through the file (not immediately) going to try to fix this.~ This was caused from `setIndex` creating a new buffer for every geometry every new message
 - add triangle scene entity to storybook test

Fixes #4318

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      entity: undefined,
+    });
+  }
+
+  private _ensureCapacity(triCount: number) {
+    if (triCount > this._triangleMeshes.length) {
+      for (let i = this._triangleMeshes.length - 1; i < triCount; i++) {
+        this._triangleMeshes.push(makeTriangleMesh());
+      }
+    }
```
Comment (jhurliman) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts:38]: Minor tweak, but you can reduce the two branches down to one using a while loop.

```
while (triCount > this._triangleMeshes.length) {
  this._triangleMeshes.push(makeTriangleMesh());
}
```

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            material.vertexColors = true;
+          }
+          geometry.attributes.color!.needsUpdate = true;
+        }
+
+        geometry.attributes.position!.needsUpdate = true;
+        geometry.computeVertexNormals();
+        geometry.computeBoundingSphere();
+      }
+      geometry.setIndex(primitive.indices);
```
Comment (jhurliman) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts:117]: Is `primitive.indices` required, or do we support unindexed triangle lists?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+      if (singleColor) {
+        material.vertexColors = false;
+        material.transparent = singleColor.a < 1.0;
+        material.depthWrite = singleColor.a >= 1.0;
```
Comment (jhurliman) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts:122]: Changing transparency/depthWrite on a material requires setting material.needsUpdate = true

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      this.add(mesh);
+      triMeshIdx++;
+    }
+  }
+
+  public override dispose(): void {
+    for (const mesh of this._triangleMeshes) {
+      mesh.geometry.dispose();
+      mesh.material.dispose();
+    }
```
Comment (jhurliman) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts:147]: I think dispose() should clear out `_triangleMeshes` as well.

**Review by jhurliman — APPROVED** [review]

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts (hunk 2 above)
Comment (snosenzo) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableTriangles.ts]: ah good call, I overlooked this initially

File: packages/studio-base/src/panels/ThreeDeeRender/stories/SceneEntities.stories.tsx [hunk 5]
```diff
@@ -198,7 +220,33 @@ function makeStoryScene({
             },
           ]),
 
-          triangles: [],
+          triangles: [
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/stories/SceneEntities.stories.tsx:223]: Maybe make these a bit bigger so they are more visible in the story? I think you also need to change the camera position and/or move the icospheres to a different place so they aren't cut off in storybook.

**Review by jtbandes — APPROVED** [review]

**Comment by snosenzo** [comment]
have confirmed that the freeze no longer occurs with the optimizations to setting the geometry indices. I've also confirmed that the meshes can properly change their indices and non-indexed shapes still work

**Pull request merged**